File: ios/Dottie/Sources/Engine/ChipSynth.swift
import Foundation

/// Port von core/.../ChipSynth.kt: Chiptune-Soundeffekte im NES-Stil,
/// Rechteckwellen und Rauschen, zur Laufzeit erzeugt — keine Audio-Assets.
/// Welche Frequenzen und Hüllkurven ein Klang hat, steht seit den
/// Ton-Sets in `SoundBank` (Port von SoundSet.kt) — hier steht nur noch,
/// wie daraus Samples werden.
///
/// Alle Funktionen liefern Mono-Samples im Bereich [-1, 1] bei
/// `sampleRate` Hz. Anders als auf Android (SoundPool mit Abspielrate)
/// werden die Tonhöhen-Varianten hier direkt beim Rendern erzeugt:
/// Abspielrate r entspricht Frequenz * r und Dauer / r (und Decay * r) —
/// das Ergebnis klingt identisch, braucht aber kein Varispeed-Setup.
enum ChipSynth {

    static let sampleRate: Int = 22050

    /// Attack-Rampe (~1,5 ms) gegen Knackser am Ton-Anfang.
    private static let attackSamples: Int = 32

    /// Lineares Fade-Out (~3 ms) gegen Knackser am Ton-Ende.
    private static let fadeOutSamples: Int = 64

    /// Rechteckwelle mit fester Frequenz. decay = Abklingrate pro Sekunde.
    static func square(
        freqHz: Float,
        seconds: Float,
        volume: Float = 0.4,
        decay: Float = 14,
        duty: Float = 0.5
    ) -> [Float] {
        return render(seconds: seconds, volume: volume, decay: decay, shape: .puls) { _ in
            (freqHz, duty)
        }
    }

    /// Dreieckwelle mit fester Frequenz — dieselbe Hüllkurve wie `square`,
    /// nur eine weichere Form. Keine Pulsbreite: Ein Dreieck hat keine,
    /// und ein Parameter, den niemand liest, wäre eine Lüge.
    static func triangle(
        freqHz: Float,
        seconds: Float,
        volume: Float = 0.4,
        decay: Float = 14
    ) -> [Float] {
        return render(seconds: seconds, volume: volume, decay: decay, shape: .dreieck) { _ in
            (freqHz, 0.5)
        }
    }

    /// Welle, deren Frequenz linear von fromHz nach toHz gleitet.
    static func sweep(
        fromHz: Float,
        toHz: Float,
        seconds: Float,
        volume: Float = 0.4,
        decay: Float = 5,
        wave: SoundBank.Wave = .puls
    ) -> [Float] {
        return render(seconds: seconds, volume: volume, decay: decay, shape: wave) { progress in
            (fromHz + (toHz - fromHz) * progress, 0.5)
        }
    }

    /// Rausch-Burst (deterministischer Seed wie im Original: Random(42)).
    static func noise(seconds: Float, volume: Float = 0.3, decay: Float = 18) -> [Float] {
        let random = KotlinRandom(intSeed: 42)
        let n = Int(seconds * Float(sampleRate))
        var out = [Float](repeating: 0, count: n)
        for i in 0..<n {
            let t = Float(i) / Float(sampleRate)
            out[i] = (random.nextFloat() * 2 - 1) * volume * envelope(index: i, total: n, t: t, decay: decay)
        }
        return out
    }

    /// Hängt mehrere Klänge nahtlos aneinander.
    static func concat(_ parts: [Float]...) -> [Float] {
        var out: [Float] = []
        out.reserveCapacity(parts.reduce(0) { $0 + $1.count })
        for part in parts {
            out.append(contentsOf: part)
        }
        return out
    }

    /// Mischt zwei Klänge übereinander (Summe, hart auf [-1, 1] begrenzt).
    static func mix(_ a: [Float], _ b: [Float]) -> [Float] {
        let n = max(a.count, b.count)
        var out = [Float](repeating: 0, count: n)
        for i in 0..<n {
            let sum = (i < a.count ? a[i] : 0) + (i < b.count ? b[i] : 0)
            out[i] = min(max(sum, -1), 1)
        }
        return out
    }

    /// Abspielrate für den Treffer-Blip: klettert innerhalb jeder 5er-Stufe
    /// eine Pentatonik hinauf (0, 2, 4, 7, 9 Halbtöne). pow wie Kotlins
    /// `Float.pow` in Double gerechnet und auf Float gerundet.
    static func hitRate(score: Int) -> Float {
        let pentatonic: [Int] = [0, 2, 4, 7, 9]
        let semitones = pentatonic[((score % 5) + 5) % 5]
        let exponent = Float(semitones) / 12
        return Float(pow(2.0, Double(exponent)))
    }

    /// Abspielrate für den Perfekt-Sound: Jede Serien-Stufe hebt den
    /// Münz-Sound um zwei Halbtöne.
    static func perfectRate(streak: Int) -> Float {
        let step = min(max(streak - 1, 0), 4) * 2
        let exponent = Float(step) / 12
        return Float(pow(2.0, Double(exponent)))
    }

    // MARK: - Fertige Effekte

    /// Ein Ereignis-Klang aus der Tabelle (`SoundBank`): Töne
    /// hintereinander, Rauschen darüber. Die einzige Stelle, an der aus
    /// einer `Voice` Samples werden — Kotlin und JavaScript tun exakt
    /// dasselbe, damit ein neues Ton-Set nirgends nachgebaut werden muss.
    ///
    /// `rate` ersetzt die SoundPool-Abspielrate von Android: Frequenz mal
    /// r, Dauer durch r, Abklingrate mal r — das klingt gleich, braucht
    /// aber kein Varispeed-Setup. Das Rauschen bleibt davon unberührt:
    /// Ein Aufprall hat keine Tonhöhe.
    static func render(_ voice: SoundBank.Voice, rate: Float = 1) -> [Float] {
        var tones: [Float] = []
        for tone in voice.tones {
            let part: [Float]
            if tone.fromHz != tone.toHz {
                part = sweep(
                    fromHz: tone.fromHz * rate,
                    toHz: tone.toHz * rate,
                    seconds: tone.seconds / rate,
                    volume: tone.volume,
                    decay: tone.decay * rate,
                    wave: tone.wave
                )
            } else if tone.wave == .dreieck {
                part = triangle(
                    freqHz: tone.fromHz * rate,
                    seconds: tone.seconds / rate,
                    volume: tone.volume,
                    decay: tone.decay * rate
                )
            } else {
                part = square(
                    freqHz: tone.fromHz * rate,
                    seconds: tone.seconds / rate,
                    volume: tone.volume,
                    decay: tone.decay * rate,
                    duty: tone.duty
                )
            }
            tones.append(contentsOf: part)
        }
        guard let rausch = voice.noise else {
            return tones
        }
        return mix(
            tones,
            noise(seconds: rausch.seconds, volume: rausch.volume, decay: rausch.decay)
        )
    }

    /// Alle Klänge eines Ton-Sets, benannt wie `SoundEvent` — die Quelle
    /// für GameAudio.
    static func effects(_ set: SoundSetId) -> [String: [Float]] {
        var out: [String: [Float]] = [:]
        for event in SoundEvent.allCases {
            out[event.rawValue] = render(SoundBank.voice(set, event))
        }
        return out
    }

    // MARK: - Rendering

    /// Rendert eine Welle; `voice` liefert pro Fortschritt Frequenz und
    /// Pulsbreite, `shape` die Form. Beide Formen teilen sich Phasenlauf
    /// und Hüllkurve — nur die eine Zeile, die aus der Phase einen Wert
    /// macht, unterscheidet sie.
    private static func render(
        seconds: Float,
        volume: Float,
        decay: Float,
        shape: SoundBank.Wave,
        voice: (Float) -> (Float, Float)
    ) -> [Float] {
        let n = Int(seconds * Float(sampleRate))
        var out = [Float](repeating: 0, count: n)
        var phase: Float = 0
        for i in 0..<n {
            let t = Float(i) / Float(sampleRate)
            let progress = n > 1 ? Float(i) / Float(n - 1) : 0
            let (freq, duty) = voice(progress)
            let wave: Float
            switch shape {
            case .puls:
                // Unverändert gegenüber der Fassung vor der Dreieckwelle.
                wave = phase < duty ? 1 : -1
            case .dreieck:
                // Um eine Viertelperiode verschoben, damit der Ton im
                // Nulldurchgang beginnt und steigt.
                let q = (phase + 0.25).truncatingRemainder(dividingBy: 1)
                wave = 1 - 4 * abs(q - 0.5)
            }
            out[i] = wave * volume * envelope(index: i, total: n, t: t, decay: decay)
            phase += freq / Float(sampleRate)
            if phase >= 1 {
                phase -= 1
            }
        }
        return out
    }

    /// Attack-Rampe, exponentielles Abklingen und End-Fade in einem.
    /// exp wie Kotlin/JVM in Double gerechnet und auf Float gerundet.
    private static func envelope(index: Int, total: Int, t: Float, decay: Float) -> Float {
        let attack: Float = index < attackSamples
            ? Float(index) / Float(attackSamples)
            : 1
        let remaining = total - index
        let fadeOut: Float = remaining < fadeOutSamples
            ? Float(remaining) / Float(fadeOutSamples)
            : 1
        return attack * fadeOut * Float(exp(Double(-decay * t)))
    }
}

File: core/src/main/kotlin/de/robinrehbein/punkt/game/ChipSynth.kt
package de.robinrehbein.punkt.game

import kotlin.math.abs
import kotlin.math.exp
import kotlin.math.pow
import kotlin.random.Random

/**
 * Die Wellenformen, die der Baukasten kennt.
 *
 * Mehr als zwei sind es bewusst nicht. Eine Sägezahnwelle klänge im
 * NES-Rahmen wie ein Fremdkörper — der Chip, dem dieses Spiel seinen
 * Klang schuldet, hatte zwei Pulskanäle, einen Dreieckkanal und
 * Rauschen, aber keine Säge. Wer eine dritte Form hinzufügt, sollte
 * also erst sagen können, welches Ton-Set sie braucht.
 */
enum class Wave {
    /**
     * Rechteck mit einstellbarer Pulsbreite — hell, schneidend, der
     * Klang des Bestands. Die Pulsbreite ist hier der Charakterregler.
     */
    PULS,

    /**
     * Dreieck: nur ungerade Oberwellen, und die fallen quadratisch statt
     * linear ab. Klingt deshalb weich und flötenartig statt schneidend,
     * und trägt in hoher Lage, wo ein Rechteck sticht. Die Pulsbreite
     * bleibt hier ungelesen — ein Dreieck hat keine.
     */
    DREIECK
}

/**
 * Purer Kotlin-Synthesizer für Chiptune-Soundeffekte im NES-Stil:
 * Rechteckwellen und Rauschen, zur Laufzeit erzeugt — keine Audio-Assets.
 * Lebt in :core (kein Android nötig), damit Phone (:app, GameAudio) und
 * Uhr (:wear, WearAudio) exakt dieselben Klänge erzeugen.
 *
 * Alle Funktionen liefern Mono-Samples im Bereich [-1, 1] bei
 * [SAMPLE_RATE] Hz. Jeder Ton bekommt eine kurze Attack-Rampe und ein
 * End-Fade gegen Knackser sowie eine exponentiell abfallende Hüllkurve
 * für den typischen perkussiven "Blip"-Charakter.
 */
object ChipSynth {

    const val SAMPLE_RATE = 22050

    /** Attack-Rampe (~1,5 ms) gegen Knackser am Ton-Anfang. */
    private const val ATTACK_SAMPLES = 32

    /** Lineares Fade-Out (~3 ms) gegen Knackser am Ton-Ende. */
    private const val FADE_OUT_SAMPLES = 64

    /** Rechteckwelle mit fester Frequenz. decay = Abklingrate pro Sekunde. */
    fun square(
        freqHz: Float,
        seconds: Float,
        volume: Float = 0.4f,
        decay: Float = 14f,
        duty: Float = 0.5f
    ): FloatArray = render(seconds, volume, decay, Wave.PULS) { _ -> freqHz to duty }

    /**
     * Dreieckwelle mit fester Frequenz — dieselbe Hüllkurve wie
     * [square], nur eine weichere Form. Keine Pulsbreite: Ein Dreieck
     * hat keine, und ein Parameter, den niemand liest, wäre eine Lüge
     * in der Signatur.
     */
    fun triangle(
        freqHz: Float,
        seconds: Float,
        volume: Float = 0.4f,
        decay: Float = 14f
    ): FloatArray = render(seconds, volume, decay, Wave.DREIECK) { _ -> freqHz to 0.5f }

    /** Welle, deren Frequenz linear von [fromHz] nach [toHz] gleitet. */
    fun sweep(
        fromHz: Float,
        toHz: Float,
        seconds: Float,
        volume: Float = 0.4f,
        decay: Float = 5f,
        wave: Wave = Wave.PULS
    ): FloatArray = render(seconds, volume, decay, wave) { progress ->
        (fromHz + (toHz - fromHz) * progress) to 0.5f
    }

    /** Rausch-Burst (deterministischer Seed, damit die WAVs stabil bleiben). */
    fun noise(seconds: Float, volume: Float = 0.3f, decay: Float = 18f): FloatArray {
        val random = Random(42)
        val n = (seconds * SAMPLE_RATE).toInt()
        val out = FloatArray(n)
        for (i in 0 until n) {
            val t = i.toFloat() / SAMPLE_RATE
            out[i] = (random.nextFloat() * 2f - 1f) * volume * envelope(i, n, t, decay)
        }
        return out
    }

    /** Hängt mehrere Klänge nahtlos aneinander. */
    fun concat(vararg parts: FloatArray): FloatArray {
        val out = FloatArray(parts.sumOf { it.size })
        var offset = 0
        for (part in parts) {
            part.copyInto(out, offset)
            offset += part.size
        }
        return out
    }

    /** Mischt zwei Klänge übereinander (Summe, hart auf [-1, 1] begrenzt). */
    fun mix(a: FloatArray, b: FloatArray): FloatArray {
        val out = FloatArray(maxOf(a.size, b.size))
        for (i in out.indices) {
            val sum = (if (i < a.size) a[i] else 0f) + (if (i < b.size) b[i] else 0f)
            out[i] = sum.coerceIn(-1f, 1f)
        }
        return out
    }

    /**
     * Abspielrate für den Treffer-Blip: klettert innerhalb jeder 5er-Stufe
     * eine Pentatonik hinauf (0, 2, 4, 7, 9 Halbtöne) — jeder Lauf spielt
     * so seine eigene kleine Melodie, mit jeder Himmelsstufe geht es
     * wieder von vorn los. Bleibt im SoundPool-Rahmen [0.5, 2.0].
     */
    fun hitRate(score: Int): Float {
        val pentatonic = intArrayOf(0, 2, 4, 7, 9)
        val semitones = pentatonic[((score % 5) + 5) % 5]
        return 2f.pow(semitones / 12f)
    }

    /**
     * Abspielrate für den Perfekt-Sound: Jede Serien-Stufe hebt den
     * Münz-Sound um zwei Halbtöne — die Serie ist hörbar, ohne dass es
     * einen eigenen Zähler auf dem Bildschirm braucht.
     */
    fun perfectRate(streak: Int): Float =
        2f.pow(((streak - 1).coerceIn(0, 4) * 2) / 12f)

    /**
     * Ein Ereignis-Klang aus der Tabelle: Töne hintereinander, Rauschen
     * darüber. Die einzige Stelle, an der aus [Voice] Samples werden —
     * die drei Handports tun exakt dasselbe, damit ein neues Ton-Set
     * nirgends nachgebaut werden muss.
     */
    fun render(voice: Voice): FloatArray {
        val tones = concat(
            *voice.tones.map { t ->
                when {
                    t.fromHz != t.toHz ->
                        sweep(t.fromHz, t.toHz, t.seconds, t.volume, t.decay, t.wave)
                    t.wave == Wave.DREIECK ->
                        triangle(t.fromHz, t.seconds, volume = t.volume, decay = t.decay)
                    else ->
                        square(t.fromHz, t.seconds, volume = t.volume, decay = t.decay, duty = t.duty)
                }
            }.toTypedArray()
        )
        val rausch = voice.noise ?: return tones
        return mix(tones, noise(rausch.seconds, volume = rausch.volume, decay = rausch.decay))
    }

    /**
     * Alle Spiel-Sounds eines Ton-Sets, benannt — die Quelle für
     * GameAudio. Die Schlüssel sind die Ereignisnamen in Kleinschrift
     * ("hit", "perfect", …): Sie stehen so in den Cache-Dateien und in
     * den Abspiel-Schichten aller Ports.
     */
    fun effects(set: SoundSetId): Map<String, FloatArray> =
        SoundEvent.entries.associate { event ->
            event.name.lowercase() to render(SoundBank.voice(set, event))
        }

    /** Das Standard-Set — der Bestand, für Aufrufer ohne eigene Wahl. */
    fun effects(): Map<String, FloatArray> = effects(SoundSetId.KLASSIK)

    /** Verpackt Samples als 16-Bit-Mono-WAV (Little Endian, 44-Byte-Header). */
    fun toWav(samples: FloatArray): ByteArray {
        val dataSize = samples.size * 2
        val out = ByteArray(44 + dataSize)
        var p = 0
        fun bytes(value: String) { for (c in value) out[p++] = c.code.toByte() }
        fun le32(value: Int) {
            out[p++] = (value and 0xFF).toByte()
            out[p++] = ((value shr 8) and 0xFF).toByte()
            out[p++] = ((value shr 16) and 0xFF).toByte()
            out[p++] = ((value shr 24) and 0xFF).toByte()
        }
        fun le16(value: Int) {
            out[p++] = (value and 0xFF).toByte()
            out[p++] = ((value shr 8) and 0xFF).toByte()
        }

        bytes("RIFF"); le32(36 + dataSize); bytes("WAVE")
        bytes("fmt "); le32(16)
        le16(1)                    // PCM
        le16(1)                    // Mono
        le32(SAMPLE_RATE)
        le32(SAMPLE_RATE * 2)      // Byte-Rate
        le16(2)                    // Block Align
        le16(16)                   // Bits pro Sample
        bytes("data"); le32(dataSize)
        for (sample in samples) {
            val s = (sample.coerceIn(-1f, 1f) * Short.MAX_VALUE).toInt()
            le16(s and 0xFFFF)
        }
        return out
    }

    /**
     * Rendert eine Welle; [voice] liefert pro Fortschritt Frequenz und
     * Pulsbreite, [shape] die Form.
     *
     * Beide Formen teilen sich Phasenlauf und Hüllkurve — nur die eine
     * Zeile, die aus der Phase einen Wert macht, unterscheidet sie. Das
     * ist der Grund, warum ein zweites Ton-Set kein zweiter Synthesizer
     * ist.
     */
    private inline fun render(
        seconds: Float,
        volume: Float,
        decay: Float,
        shape: Wave,
        voice: (progress: Float) -> Pair<Float, Float>
    ): FloatArray {
        val n = (seconds * SAMPLE_RATE).toInt()
        val out = FloatArray(n)
        var phase = 0f
        for (i in 0 until n) {
            val t = i.toFloat() / SAMPLE_RATE
            val (freq, duty) = voice(if (n > 1) i.toFloat() / (n - 1) else 0f)
            val wave = when (shape) {
                // Unverändert gegenüber der Fassung vor der Dreieckwelle:
                // Der Bestand darf sich nicht um ein Sample verschieben.
                Wave.PULS -> if (phase < duty) 1f else -1f
                // Um eine Viertelperiode verschoben, damit der Ton im
                // Nulldurchgang beginnt und steigt. Ohne die Verschiebung
                // stünde bei Phase 0 der Scheitel, und die Attack-Rampe
                // arbeitete gegen einen Sprung von 0 auf 1.
                Wave.DREIECK -> {
                    val q = (phase + 0.25f) % 1f
                    1f - 4f * abs(q - 0.5f)
                }
            }
            out[i] = wave * volume * envelope(i, n, t, decay)
            phase += freq / SAMPLE_RATE
            if (phase >= 1f) phase -= 1f
        }
        return out
    }

    /** Attack-Rampe, exponentielles Abklingen und End-Fade in einem. */
    private fun envelope(index: Int, total: Int, t: Float, decay: Float): Float {
        val attack = if (index < ATTACK_SAMPLES) index.toFloat() / ATTACK_SAMPLES else 1f
        val remaining = total - index
        val fadeOut = if (remaining < FADE_OUT_SAMPLES) remaining.toFloat() / FADE_OUT_SAMPLES else 1f
        return attack * fadeOut * exp(-decay * t)
    }
}
